File: partiql-ast/src/main/java/org/partiql/ast/v1/OrderBy.java
package org.partiql.ast.v1;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * TODO docs, equals, hashcode
 */
@Builder(builderClassName = "Builder")
@EqualsAndHashCode(callSuper = false)
public class OrderBy extends AstNode {
    @NotNull
    public final List<Sort> sorts;

    public OrderBy(@NotNull List<Sort> sorts) {
        this.sorts = sorts;
    }

    @NotNull
    @Override
    public Collection<AstNode> children() {
        return new ArrayList<>(sorts);
    }

    @Override
    public <R, C> R accept(@NotNull AstVisitor<R, C> visitor, C ctx) {
        return visitor.visitOrderBy(this, ctx);
    }
}

File: partiql-ast/src/main/kotlin/org/partiql/ast/v1/OrderBy.kt
package org.partiql.ast.v1

/**
 * TODO docs, equals, hashcode
 */
public class OrderBy(
    @JvmField
    public var sorts: List<Sort>,
) : AstNode() {
    public override fun children(): Collection<AstNode> {
        val kids = mutableListOf<AstNode?>()
        kids.addAll(sorts)
        return kids.filterNotNull()
    }

    public override fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R =
        visitor.visitOrderBy(this, ctx)
}
